platforms/jvm/plugins-application/src/main/java/org/gradle/api/plugins/ApplicationPlugin.java

```diff
@@ -91,6 +92,14 @@ private void configureJavaCompileTask(TaskProvider<JavaCompile> javaCompile, Jav
         javaCompile.configure(j -> j.getOptions().getJavaModuleMainClass().convention(pluginExtension.getMainClass()));
     }
 
+    private void configureJarTask(TaskProvider<Jar> jar, JavaApplication pluginExtension) {
+        jar.configure(j -> {
+            if (pluginExtension.getMainClass().isPresent()) {
+                j.getManifest().getAttributes().put("Main-Class", pluginExtension.getMainClass().get());
+            }
```

Comment:
This still has ordering issues because `mainClass` could be changed after this runs.

The underlying problem is that the manifest's attributes do not accept Providers, ideally you'd be able to write something like:
```
jar.manifest.attributes.put("Main-Class", pluginExtension.mainClass)
```
And everything would be lazy.

Until then, we need to consider using `afterEvaluate` or see if there's something simple we can add to Manifest or Attributes to support Provider.

Reply:
@big-guy, I've updated the PR with wrapping the task configuration in `afterEvaluate`.